### ISSUE TITLE
What does it take to make StackStorm Available and Scalable?

### ISSUE THREAD

**Issue by manasdk** [StackStorm/discussions#11]
## Preface

Needed for v1.0 or so we believe.

This discussion does not capture hows just the problems so that we can choose which ones need to be solved. It does not preclude alternates to provided proposals just some food for thought. I could've written out concrete solutions but then I would presume that I understand all the requirements and the whole problem space - which I do not.
## Some definitions

Reliable - Does what it says it does within specification.
Available - Probability that system is operational at any given time.
Scalable - Can be made to do more work provided more resources are given.

Reliable is basically bug free so we are talking only about `Availability` and `Scalability` of st2. Glad that is clear :)
## Assumptions
1. st2 is expected to run on multiple boxes
2. should also work in a single box deploy
3. all content is available on all content consumer nodes 
4. content registration is managed on the central node (?). Not sure how microservice/containerized deploy would work.
## Lets worry about availability first and scale later

Scale matters but st2 intrinsically is capable of managing some load. Lets just say we have 99 problems but scale ain't one(yet).

On the other hand we don't have any understanding of availability of st2.
## st2 availability

Since st2 is basically a set of disjoint process that co-operatively achieve work it follows that we achieve high-availability of each process independently. Generally lets go with `availability through redundancy`
### sensors

SensorContainer is a manager process that spawns off each sensor process independently. A SensorManager will run each registered sensor except if it is running in a single sensor mode. Since the SensorManager decides which sensor to run it is not possible to shard sensors and restart of a single sensor leads to a restart of all sensors.
#### Proposal/Options
1. Some form of active-passive sharding of sensors
2. Ability to turn sensors on and off live
3. Allow a sensor to require multiple instances of itself in-case the sensor knows how to shard/load-balance itself. [May need some anti-affinity controls]
4. Ability to restart crashed sensors. Perhaps even a policy governing auto-restarts.
### RulesEngine

We can run RulesEngine in an Active-Active mode expect that timer thread would be an issue.
#### Proposal/Options

Master or sharded mode for timers. Master-mode is easier rather than sharding of timers as sharded timers would've issues with recovery.
### ActionRunner

Works well in a active-active availability mode.
### ResultsTracker

Does not support active-active.
#### Proposal
1. Restart on crash
### Notifier + Scheduler

Does not support active-active.
#### Proposal
1. Restart on crash
### API

Supports active-active and can be put behind a load-balancer.
## Reporting state

Lets face it st2 will not always be available therefore we must keep heart-beating out from all processes. `Whenever, a HB is missing we know that some st2 process was down.`
### Proposal
1. Provide stackstorm some monitoring tool to report into and each process will do the needful.
2. A service registry to co-relate nodes and roles.
3. Somewhere down the line an API to easily gather this information.
## TLDR
1. sensorcontainer and rulesengine to support some form of active-active or active-passive modes. 
2. resultstracker and notifier+scheduler do not operate in a redundant way yet. We report failure and at most restart.
### Monitoring & Reporting
1. Post to graphite via statsd client (+ve nice visuals)
2. Nagios/sensu monitoring
### Serivce registry
1. consul
### Recovery
1. Watchdog - is there a nice distributed/reporting watchdog we can use?

**Comment by armab** [comment]
This Discussion issue was moved from the private Discussions to make it available to public community as it's still actual today.
